homeassistant/components/proximity/coordinator.py

```diff
@@ -0,0 +1,268 @@
+
+    async def _async_update_data(self) -> ProximityData:
+        """Calculate Proximity data."""
+        if self.state_change_data is None or self.state_change_data.new_state is None:
+            return self.data
+
+        entity_name = self.state_change_data.new_state.name
+        devices_to_calculate = False
+        devices_in_zone = ""
+
+        zone_state = self.hass.states.get(f"zone.{self.proximity_zone}")
+        proximity_latitude = (
+            zone_state.attributes.get(ATTR_LATITUDE) if zone_state else None
+        )
+        proximity_longitude = (
+            zone_state.attributes.get(ATTR_LONGITUDE) if zone_state else None
+        )
+
+        # Check for devices in the monitored zone.
+        for device in self.proximity_devices:
+            if (device_state := self.hass.states.get(device)) is None:
+                devices_to_calculate = True
+                continue
+
+            if device_state.state not in self.ignored_zones:
+                devices_to_calculate = True
+
+            # Check the location of all devices.
+            if (device_state.state).lower() == (self.proximity_zone).lower():
+                device_friendly = device_state.name
+                if devices_in_zone != "":
+                    devices_in_zone = f"{devices_in_zone}, "
+                devices_in_zone = devices_in_zone + device_friendly
+
+        # No-one to track so reset the entity.
+        if not devices_to_calculate:
+            _LOGGER.debug("no devices_to_calculate -> abort")
+            return {
+                "dist_to_zone": DEFAULT_DIST_TO_ZONE,
+                "dir_of_travel": DEFAULT_DIR_OF_TRAVEL,
+                "nearest": DEFAULT_NEAREST,
+            }
+
+        # At least one device is in the monitored zone so update the entity.
+        if devices_in_zone != "":
+            _LOGGER.debug("at least on device is in zone -> arrived")
+            return {
+                "dist_to_zone": 0,
+                "dir_of_travel": "arrived",
+                "nearest": devices_in_zone,
+            }
+
+        # Collect distances to the zone for all devices.
+        distances_to_zone: dict[str, float] = {}
+        for device in self.proximity_devices:
+            # Ignore devices in an ignored zone.
+            device_state = self.hass.states.get(device)
+            if not device_state or device_state.state in self.ignored_zones:
+                continue
+
+            # Ignore devices if proximity cannot be calculated.
+            if "latitude" not in device_state.attributes:
+                continue
+
+            # Calculate the distance to the proximity zone.
+            proximity = distance(
+                proximity_latitude,
+                proximity_longitude,
+                device_state.attributes[ATTR_LATITUDE],
+                device_state.attributes[ATTR_LONGITUDE],
+            )
+
+            # Add the device and distance to a dictionary.
+            if not proximity:
```

Comment:
Can the proximity be 0? Should we handle that the same as `None`?

Reply:
2c9234e2b93265f91bcc3e27a2e925996b258c1e